lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -37,196 +35,206 @@ define((require, exports, module) ->
 
     # Check if current word is in corrections list.
     # Call showPopup if it's present, do nothing otherwise.
-    tryPopup = ->
-      word = extractWord()
+    newPopup = ->
+      word = extractWord(editor)
       spellChecker = new SpellChecker()
       correctionsItem = spellChecker.getCorrections(word)
       if correctionsItem
-        showPopup(convertCorrectionList(correctionsItem))
+        PopupManager.show(convertCorrectionList(correctionsItem))
       return
 
-    # Bind a shortcut to tryPopup callback.
+    # Bind newPopup to Alt-Enter editor shortcut
     editor.commands.addCommand({
-      name: "test",
+      name: "newPopup",
       bindKey: "Alt-Enter",
-      exec: tryPopup
+      exec: newPopup
     })
-
     return
 
+
+  # Get the word under the cursor.
+  # @param {Editor} editor: editor object
+  # @return {String}
+  extractWord = (editor) ->
+    session = editor.session
+    row = editor.getCursorPosition().row
+    wordRange = getCurrentWordRange(editor)
+    # getAWordRange returns start and end positions with a trailing
+    # whitespace at the end (if there's a one), that's why replace is used
+    return session.getTextRange(wordRange).replace(/\s\s*$/, '')
+
+  # Returns Range object that describes current word position.
+  # @param {Editor} editor: editor object
+  # @return {Range}: current word range.
+  getCurrentWordRange = (editor) ->
+    session = editor.session
+    row = editor.getCursorPosition().row
+    col = editor.getCursorPosition().column
+    return session.getAWordRange(row, col)
+
+  # Convert an array of string to popup eligible structure.
+  # @param {Array} corrections: array of strings with substitution options.
+  # @return {Array}: array of jsons, actually.
+  convertCorrectionList = (corrections) ->
+    options = []
+    options.push({caption: item, value: item}) for item in corrections
+    return options
+
+
+  # TODO: incorrect token without correction list.
+  # Object contains init, show and detach functions for popup and some
+  # routines for inner use.
+  PopupManager =
+    # Bindings for editor keys with PopupManager functions.
+    commands:
+      "Esc": ->
+        PopupManager.detach()
+      "Up": ->
+        PopupManager.goTo("up")
+      "Down": ->
+        PopupManager.goTo("down")
+      "Return": ->
+        PopupManager.insertCorrection()
+
+    # Initial setup for popup.
+    # @param {Editor} editor: editor object.
+    init: (editor) ->
+      @editor = editor
+
+      HashHandler = require("ace/keyboard/hash_handler").HashHandler
+      AcePopup = require("ace/autocomplete/popup").AcePopup
+
+      @popup = new AcePopup(document.body)
+      @popup.setTheme(editor.getTheme())
+      @popup.setFontSize(editor.getFontSize())
+      @popup.on("click", (e) =>
+        @insertCorrection()
+        e.stop()
+      )
+
+      @session = editor.session
+
+      @keyboard = new HashHandler()
+      @keyboard.bindKeys(@commands)
+
+      @detach if editor?
```

Comment:
? Why do you need to detach in the init, why argument braces are missing and how come you expect editor to be null after you have accessed its methods a number of times?